app/src/main/cpp/skyline/kernel/scheduler.cpp

```diff
@@ -0,0 +1,367 @@
+        }
+
+        thread->averageTimeslice = (thread->averageTimeslice / 4) + (3 * (util::GetTimeTicks() - thread->timesliceStart / 4));
+
+        if (cooperative && thread->isPreempted) {
+            // If a preemptive thread did a cooperative yield then we need to disarm the preemptive timer
+            struct itimerspec spec{};
+            timer_settime(thread->preemptionTimer, 0, &spec, nullptr);
+        }
+
+        thread->isPreempted = false;
+        thread->pendingYield = false;
+        thread->forceYield = false;
+    }
+
+    void Scheduler::RemoveThread() {
+        auto &thread{state.thread};
+        auto &core{cores.at(thread->coreId)};
+        {
+            std::unique_lock lock(core.mutex);
+            auto it{std::find(core.queue.begin(), core.queue.end(), thread)};
+            if (it != core.queue.end()) {
+                it = core.queue.erase(it);
+                if (it == core.queue.begin()) {
+                    // We need to update the averageTimeslice accordingly, if we've been unscheduled by this
+                    if (thread->timesliceStart)
+                        thread->averageTimeslice = (thread->averageTimeslice / 4) + (3 * (util::GetTimeTicks() - thread->timesliceStart / 4));
+
+                    if (it != core.queue.end())
+                        (*it)->wakeCondition.notify_one(); // We need to wake the thread at the front of the queue, if we were at the front previously
+                }
+            }
+        }
+
+        if (thread->isPreempted) {
+            struct itimerspec spec{};
+            timer_settime(thread->preemptionTimer, 0, &spec, nullptr);
+            thread->isPreempted = false;
+        }
+
+        YieldPending = false;
+    }
+
+    void Scheduler::UpdatePriority(const std::shared_ptr<type::KThread> &thread) {
+        std::lock_guard migrationLock(thread->coreMigrationMutex);
+        auto *core{&cores.at(thread->coreId)};
+        std::unique_lock coreLock(core->mutex);
+
+        auto currentIt{std::find(core->queue.begin(), core->queue.end(), thread)};
+        if (currentIt == core->queue.end() || currentIt == core->queue.begin())
+            // If the thread isn't in the queue then the new priority will be handled automatically on insertion
+            return;
+        if (currentIt == core->queue.begin()) {
+            // Alternatively, if it's currently running then we'd just want to cause it to yield if it's priority is lower than the the thread behind it
+            auto nextIt{std::next(currentIt)};
+            if (nextIt != core->queue.end() && (*nextIt)->priority < thread->priority) {
+                if (!thread->pendingYield) {
+                    thread->SendSignal(YieldSignal);
+                    thread->pendingYield = true;
+                }
+            } else if (!thread->isPreempted && thread->priority == core->preemptionPriority) {
```

Comment:
I think we should have a helper function for the timer setting, it's repeated a bunch

Reply:
There's a lot of tiny snippets like this (3 instances of `SendSignal(YieldSIgnal)`), I think it's alright to duplicate them due to them being so minimal.

Reply:
Im meaning the one below, but I disagree, it adds some clarity as to what they are too.

Reply:
I am aware, that was an example of another snippet which is repeated.

Reply:
but do you agree or not?

Reply:
I do not as I think there's far too many of the 2-3 line snippets and it'll just clutter things up further.

Reply:
I think this isn't especially clear either with what it does too

Reply:
I still think this point is valid, it deserves its own helper.

Reply:
I still think this point is valid, it's ugly c boilerplate with no need to be repeated, it would also allow trying different timers easier and logging stuff